hw/misc/hermes.c

```diff
@@ -383,12 +476,103 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
                 addr);
 }
 
+static inline void mark_exec_finished(HermesState *hermes, int engine)
+{
+    bar0_init.cmdctrl[engine].ehcmddone = HERMES_CMD_FINISHED;
+    trace_hermes_msix_notify(H2C_CHANNELS + C2H_CHANNELS + engine);
+    msix_notify(PCI_DEVICE(hermes), H2C_CHANNELS + C2H_CHANNELS + engine);
+}
+
+static void hermes_exec(HermesState *hermes, int engine)
+{
+    struct hermes_cmd *command = &bar0_init.commands[engine];
+
+    void *bar4_base = memory_region_get_ram_ptr(&hermes->bar4_mem_reg);
+
+    char *errmsg;
+    int32_t rv;
+    uint64_t ret;
+    bool elf;
+
+    unsigned char prog_slot = command->req.run_prog.prog_slot;
+    unsigned char data_slot = command->req.run_prog.data_slot;
+
+    void *program_base = bar4_base + bar0_init.ehpsoff +
+                         (prog_slot * bar0_init.ehpssze);
+    void *data_base = bar4_base + bar0_init.ehdsoff +
+                      (data_slot * bar0_init.ehdssze);
+
+    if (prog_slot >= bar0_init.ehpslot) {
+        command->res.status = HERMES_STATUS_INVALID_PROG_SLOT;
+        mark_exec_finished(hermes, engine);
+        return;
+    }
+
+    if (data_slot >= bar0_init.ehdslot) {
+        command->res.status = HERMES_STATUS_INVALID_DATA_SLOT;
+        mark_exec_finished(hermes, engine);
+        return;
+    }
+
+    command->res.cid = command->req.cid;
+
+    if (command->req.opcode != HERMES_OPCODE_EXECUTE_SLOT) {
+        command->res.status = HERMES_STATUS_INVALID_OPCODE;
+        hermes_bar_warn_invalid_command(command->req.opcode, engine);
+        mark_exec_finished(hermes, engine);
+        return;
+    }
+
+    /* Check magic number (first 4 bytes), to see if program is in ELF format */
+    elf = bar0_init.ehpssze >= SELFMAG && !memcmp(program_base, ELFMAG, 4);
+    if (elf) {
+        rv = ubpf_load_elf(hermes->ubpf[engine].engine, program_base,
```

Comment:
We're going to have problems with this... The current ubpf_load* functions do not support loading a second program on top of the previous one.

I'm going to submit a PR to the ubpf project to address this, but in the meantime, can we add the following before trying to load?

```C
free(hermes->ubpf[engine].engine->insts);
hermes->ubpf[engine].engine->insts = NULL;
```